### ISSUE TITLE
NumPy 2.0 compatibility

### ISSUE THREAD

**Issue by qiyunzhu** [scikit-bio/scikit-bio#1964]
It seems that scikit-bio is compatible with the upcoming [NumPy 2.0](https://numpy.org/devdocs/release/2.0.0-notes.html). Following this [guideline](https://numpy.org/devdocs/numpy_2_0_migration_guide.html), I ran:

```bash
ruff check skbio --select NPY201
```

And nothing happened, meaning the check passed. However, I haven't been able to actually install NumPy 2.0 for test (probably because it is still under active development). Therefore we should keep an eye on it.

Thanks @wasade for pointing out this.

**Comment by rgommers** [comment]
You can install numpy 2.0.0rc1 from PyPI now, with `pip install numpy --pre`. It'd be great to get a confirmation that everything works well with numpy 2.x

**Comment by qiyunzhu** [comment]
@rgommers Thanks for pointing us to this!

Just tested. Got error:

```
File "h5py/_errors.pyx", line 1, in init h5py._errors
ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject
```

This is probably an issue in h5py, a dependency of scikit-bio. I saw this [discussion](https://github.com/h5py/h5py/issues/2353) in the h5py codebase. We probably need to wait until after they release a NumPy 2.0 compatible version. @wasade

**Comment by qiyunzhu** [comment]
@rgommers Update: I was able to install numpy 2.0.0rc1 and compatible versions of scipy, pandas and h5py. Then got the following error when importing scikit-bio:

```
  File "skbio/stats/__subsample.pyx", line 1, in init skbio.stats.__subsample
ImportError: numpy.core.multiarray failed to import (auto-generated because you didn't call 'numpy.import_array()' after cimporting numpy; use '<void>numpy._import_array' to disable if you are certain you don't need it).
```

We have planned to replace the `__subsample` function with a new one so this error might become irrelevant in the next release. But if you have some insights to share we will appreciate.

**Comment by rgommers** [comment]
Not too much more than what the error message says:
```
because you didn't call 'numpy.import_array()' after cimporting numpy
```
in the Cython file where this error comes from, I assume there's a `cimport numpy as cnp` statement. and no `cnp.import_array()`.

**Comment by rgommers** [comment]
`h5py` release is out now - slowly getting there:)

**Comment by qiyunzhu** [comment]
Thanks for the update! Now we are waiting for Pandas 2.2.2. Then we add [numpy 2.0 support](https://github.com/biocore/biom-format/pull/950) for the upstream package biom-format. Then we can unblock numpy 2.0 in scikit-bio.

**Comment by rgommers** [comment]
> Now we are waiting for Pandas 2.2.2.

That's up on PyPI too now, since ~25 minutes ago.

**Comment by rgommers** [comment]
Hi @qiyunzhu, I just wanted to check in on this, and ask if you plan to include support in 0.6.1 or 0.7.0 (and, if you know, roughly when)?

**Comment by qiyunzhu** [comment]
Hello @rgommers , We recently resolved the NumPy 2.0 support issue in the upstream package biom-format (see [here](https://github.com/biocore/biom-format/pull/962)). We are about to resolve this in scikit-bio in the next few days. It will be 0.6.1. Let's say it will be released in a month or so.

**Comment by rgommers** [comment]
Thanks! I'm not 100% sure that the final numpy 2.0 release will take more than a month from now, so a release before the end of May would be safer. But if you've got all the problems resolved then that's already good to know.

**Comment by jakirkham** [comment]
JFYI NumPy is planning to release 2.0 on June 16th

xref: https://github.com/numpy/numpy/issues/24300#issuecomment-2125643448

**Comment by sspola** [comment]
Hi,

I am a new user and tried to install scikit-bio today for the first time.

After a successful installation in a new environment, I got the following error on import of scikit-bio
`ImportError: numpy.core.multiarray failed to import`

It seems that some modules are compiled using NumPy 1.x and thus cannot be run in NumPy 2.0.0 (which was released yesterday and installed together with the other dependencies of scikit-bio).

Downgrading to the newest 1.x version fixed the problem.

**Comment by qiyunzhu** [comment]
@sspola Thanks for the report. We are preparing a NumPy 2.0-compatible release. It will be out very soon. At present, you can do `pip install git+https://github.com/scikit-bio/scikit-bio.git` to install the nightly build which should work with NumPy 2.0. Let me know if it works or not.

**Comment by qiyunzhu** [comment]
@sspola We have released scikit-bio 0.6.1, which is compatible with NumPy 2.0. 😊

**Comment by sspola** [comment]
@qiyunzhu that's great, thank you!
